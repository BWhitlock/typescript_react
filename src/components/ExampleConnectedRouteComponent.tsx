import * as React from "react";
import { connect } from "react-redux";
import { RouteComponentProps } from "react-router";
import { bindActionCreators, Dispatch } from "redux";
import { createExampleAction } from "../actions/example";
import { Action } from "../actions/types";
import { State } from "../types/state";

type OwnProps = {};

type StateProps = {
  flag: boolean;
};

type ActionProps = {
  createExampleAction: (flag: boolean) => Action;
};

type RouteProps = {
  message?: string;
};

export class ExampleConnectedRouteComponent extends React.Component<
  StateProps & ActionProps & OwnProps & RouteComponentProps<RouteProps>
> {
  toggleFlag = () => this.props.createExampleAction(!this.props.flag);

  render() {
    return (
      <div>
        <h2>
          {this.props.match.params.message || "No message."}
        </h2>
        <p>
          {this.props.flag ? "flag is set" : "flag is not set"}
        </p>
        <button onClick={this.toggleFlag}>Toggle the flag</button>
      </div>
    );
  }
}

const mapStateToProps = (
  state: State,
  ownProps: OwnProps
): StateProps & OwnProps => ({
  flag: state.example.flag,
  ...ownProps
});

const mapDispatchToProps = (dispatch: Dispatch<State>): ActionProps =>
  bindActionCreators({ createExampleAction }, dispatch);

export default connect<StateProps, ActionProps, OwnProps>(
  mapStateToProps,
  mapDispatchToProps
)(ExampleConnectedRouteComponent);
