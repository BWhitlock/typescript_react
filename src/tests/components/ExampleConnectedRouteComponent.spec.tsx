import { shallow } from "enzyme";
import * as React from "react";
import { ExampleConnectedRouteComponent } from "../../components/ExampleConnectedRouteComponent";
import mockRouteProps from "./utils/mockRouteProps";

describe("ExampleConnectedRouteComponent", () => {
  it("matches snapshot", () => {
    const routeProps = mockRouteProps();

    const tree = shallow(
      <ExampleConnectedRouteComponent
        flag={true}
        createExampleAction={jest.fn()}
        {...routeProps}
      />
    );
    expect(tree).toMatchSnapshot();
  });

  it("Clicking button calls createExampleAction", () => {
    const mockCreateExampleAction = jest.fn();
    const routeProps = mockRouteProps();
    const component = shallow(
      <ExampleConnectedRouteComponent
        flag={false}
        createExampleAction={mockCreateExampleAction}
        {...routeProps}
      />
    );

    expect(mockCreateExampleAction).not.toHaveBeenCalled();
    component.find("button").simulate("click");
    expect(mockCreateExampleAction).toHaveBeenCalled();
  });
});
