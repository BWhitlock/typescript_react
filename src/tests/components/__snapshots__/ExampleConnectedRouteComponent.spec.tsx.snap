// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExampleConnectedRouteComponent matches snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <h2>
        No message.
    </h2>
    <p>
        flag is set
    </p>
    <button
        onClick={[Function]}
    >
        Toggle the flag
    </button>
</div>,
  "nodes": Array [
    <div>
      <h2>
            No message.
      </h2>
      <p>
            flag is set
      </p>
      <button
            onClick={[Function]}
      >
            Toggle the flag
      </button>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ExampleConnectedRouteComponent
        createExampleAction={[Function]}
        flag={true}
        history={
                Object {
                        "action": "REPLACE",
                        "block": [Function],
                        "createHref": [Function],
                        "go": [Function],
                        "goBack": [Function],
                        "goForward": [Function],
                        "length": 0,
                        "listen": [Function],
                        "location": Object {
                          "hash": "",
                          "key": "",
                          "pathname": "",
                          "search": "",
                          "state": Object {},
                        },
                        "push": [Function],
                        "replace": [Function],
                      }
        }
        location={
                Object {
                        "hash": "",
                        "key": "",
                        "pathname": "",
                        "search": "",
                        "state": Object {},
                      }
        }
        match={
                Object {
                        "isExact": false,
                        "params": Object {},
                        "path": "",
                        "url": "",
                      }
        }
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ExampleConnectedRouteComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "createExampleAction": [Function],
          "flag": true,
          "history": Object {
            "action": "REPLACE",
            "block": [Function],
            "createHref": [Function],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "length": 0,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "key": "",
              "pathname": "",
              "search": "",
              "state": Object {},
            },
            "push": [Function],
            "replace": [Function],
          },
          "location": Object {
            "hash": "",
            "key": "",
            "pathname": "",
            "search": "",
            "state": Object {},
          },
          "match": Object {
            "isExact": false,
            "params": Object {},
            "path": "",
            "url": "",
          },
        },
        "refs": Object {},
        "state": null,
        "toggleFlag": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          <h2>
                    No message.
          </h2>
          <p>
                    flag is set
          </p>
          <button
                    onClick={[Function]}
          >
                    Toggle the flag
          </button>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          <h2>
                    No message.
          </h2>
          <p>
                    flag is set
          </p>
          <button
                    onClick={[Function]}
          >
                    Toggle the flag
          </button>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ExampleConnectedRouteComponent
    createExampleAction={[Function]}
    flag={true}
    history={
        Object {
            "action": "REPLACE",
            "block": [Function],
            "createHref": [Function],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "length": 0,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "key": "",
              "pathname": "",
              "search": "",
              "state": Object {},
            },
            "push": [Function],
            "replace": [Function],
          }
    }
    location={
        Object {
            "hash": "",
            "key": "",
            "pathname": "",
            "search": "",
            "state": Object {},
          }
    }
    match={
        Object {
            "isExact": false,
            "params": Object {},
            "path": "",
            "url": "",
          }
    }
/>,
}
`;
